src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/Frame.cs

```diff
@@ -498,63 +504,126 @@ public void OnCompleted(Func<object, Task> callback, object state)
             }
         }
 
-        protected async Task FireOnStarting()
+        protected Task FireOnStarting()
         {
-            Stack<KeyValuePair<Func<object, Task>, object>> onStarting = null;
+            Stack<KeyValuePair<Func<object, Task>, object>> onStarting;
             lock (_onStartingSync)
             {
                 onStarting = _onStarting;
                 _onStarting = null;
             }
-            if (onStarting != null)
+
+            if (onStarting == null)
             {
-                try
+                return TaskCache.CompletedTask;
+            }
+            else
+            {
+                return FireOnStartingMayAwait(onStarting);
+            }
+
+        }
+
+        private Task FireOnStartingMayAwait(Stack<KeyValuePair<Func<object, Task>, object>> onStarting)
+        {
+            try
+            {
+                foreach (var entry in onStarting)
                 {
-                    foreach (var entry in onStarting)
+                    var task = entry.Key.Invoke(entry.Value);
+                    if (!ReferenceEquals(task, TaskCache.CompletedTask))
                     {
-                        await entry.Key.Invoke(entry.Value);
+                        return FireOnStartingAwaited(task, onStarting);
```

Comment:
Was thinking it should have removed it from stack, but you are right as it is only enumerating. Will change